source/material_model/rheology/strain_dependent.cc

```diff
@@ -415,7 +494,15 @@ namespace aspect
         if  (this->simulator_is_past_initialization() && this->get_timestep_number() > 0 && in.requests_property(MaterialProperties::reaction_terms))
           {
             const double edot_ii = std::max(sqrt(std::fabs(second_invariant(deviator(in.strain_rate[i])))),min_strain_rate);
-            const double e_ii = edot_ii*this->get_timestep();
+            double e_ii = edot_ii*this->get_timestep();
+
+            // Adjusting strain values to account for strain healing without exceeding an unreasonable range
+            if (healing_mechanism != no_healing)
+              {
+                // Never heal more strain than exists
+                const double healed_strain = std::min(calculate_strain_healing(in,i), e_ii);
+                e_ii -= healed_strain;
+              }
```

Comment:
Ah, wait a second. I was confused by the name `e_ii`. It is not the accumulated strain, but the amount of strain that gets added in this timestep (`edot_ii * dt`). So it should be named `delta_e_ii`, and the amount of healed strain should not be limited to `delta_e_ii`, but to the total amount of accumulated strain plus the newly added strain (`e_ii + delta_e_ii`). Unfortunately we do not know the value of the accumulated strain here, because it depends on which strain is used for strain weakening (the if conditions below). So we need to unconditionally add the healing term here, and then check below that it does not exceed the value of the respective accumulated strain. Did I understand this correctly?

```suggestion
            double delta_e_ii = edot_ii*this->get_timestep();

            if (healing_mechanism != no_healing)
                delta_e_ii -= calculate_strain_healing(in,i);
```

Reply:
Yes, that would be right